### PR TITLE
Add taxes to undiscounted order prices

### PR DESCRIPTION
**Pull request by jakubkuc**
I want to merge this change because it is a port of #12284 but we decided to release it for Saleor `3.14` and up instead of `3.9` so it needed some extra adjustments. 

All the comments from the old PR are addressed, but It still needs a review since the base is different. 

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migrations are either absent or optimized for zero downtime
* [ ] The changes are covered by test cases

### CODE REVIEW

File: saleor/checkout/tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            Decimal(0.25),
+            False,
+            TaxedMoney(gross=Money("43.74", "USD"), net=BASE),
+        ),
+        # result should be calculated and different from price
+        (
+            TaxedMoney(gross=Money("43.74", "USD"), net=Money("36.00", "USD")),
+            Decimal(0.25),
+            False,
+            TaxedMoney(gross=Money("43.75", "USD"), net=BASE),
```
Comment (IKarbowiak) [inline saleor/checkout/tests/test_utils.py:26]: Why this gross is different than gross from line 23?

File: saleor/checkout/tests/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+)
+def test_get_taxed_undiscounted_price(price, tax_rate, prices_entered_with_tax, result):
+    result_price = get_taxed_undiscounted_price(
+        undiscounted_base_price=BASE,
+        price=price,
+        tax_rate=tax_rate,
+        prices_entered_with_tax=prices_entered_with_tax,
+    )
+
+    assert result_price == result
```
Comment (IKarbowiak) [inline saleor/checkout/tests/test_utils.py:52]: IMO, instead of hardcoding the result, it should be calculated here based on the `tax_rate`. Otherwise the test is kind of magic and hard to maintenance.

Comment (jakubkuc) [reply saleor/checkout/tests/test_utils.py:52]: It kinda is, I agree. Though I had no other Idea how to test with `parameterize` all the scenarios for this function  - it checks and takes into consideration the corner-case when `tax_app` is returning slightly different results than local saleor calculations based on `tax_rate` ([this comment](https://github.com/saleor/saleor/pull/13585/files/8368e1944d6cd2b4ad8f741962864974b5203f00#diff-2f35ab82d7f669cedfbd9a04be1c39c1a1c3ee299730d7f0470254f5cc37721fR969)) 🤔 .

Comment (IKarbowiak) [reply saleor/checkout/tests/test_utils.py:52]: Firstly this test probably should be split into two separates - one for checking when `prices_entered_with_tax` is `True`, when when it's `False`. Then as a prameters you can provide the `price` (I see that `tax_rate` is the same in all examples so there is no reason to parametrize it) and based on `tax_rate` we should calculate the `net` or `gross` amount and compare it with `result_price`

File: saleor/checkout/tests/test_utils.py (hunk 1 above)
Comment (jakubkuc) [reply saleor/checkout/tests/test_utils.py:26]: It is connected to the functionality described [here](https://github.com/saleor/saleor/pull/13585/files/8368e1944d6cd2b4ad8f741962864974b5203f00#diff-2f35ab82d7f669cedfbd9a04be1c39c1a1c3ee299730d7f0470254f5cc37721fR969) that takes into consideration that if local tax calculations for undiscounted prices should be the same, the price is always taken from the tax_app.

Comment (jakubkuc) [reply saleor/checkout/tests/test_utils.py:26]: I suppose that since I have to explain it it is not properly described in docstring and comments so I will need to fix that (unless you have a concern regarding this whole solution)

File: saleor/plugins/manager.py [hunk 3]
```diff
@@ -448,7 +448,7 @@ def calculate_checkout_line_unit_price(
             default_value * quantity,
         )
         default_taxed_value = TaxedMoney(
-            net=total_value / quantity, gross=default_value
+            net=total_value / quantity, gross=total_value / quantity
```
Comment (korycins) [inline saleor/plugins/manager.py:451]: Isn't something that can cause incorrect values on our side? Shouldn't we patch them also in previous versions?

File: saleor/order/migrations/tasks/saleor3_14.py [hunk 4]
```diff
@@ -15,3 +26,131 @@ def order_propagate_expired_at_task():
...
+    order_to_prices_with_tax_map = get_order_prices_entered_with_tax_mapping(
+        order_lines,
+        order_model=Order,
+        channel_model=Channel,
+        tax_configuration=TaxConfiguration,
+    )
+
+    lines_to_update = set()
+    for line in order_lines:
+        prices_entered_with_tax = order_to_prices_with_tax_map[line.order.id]
```
Comment (korycins) [inline saleor/order/migrations/tasks/saleor3_14.py:138]: This won't work. The prefetch_related is incorrect. It will raise:
```
ValueError: 'order_id' does not resolve to an item that supports prefetching - this is an invalid parameter to prefetch_related()
```
Have you tested it locally?

Comment (korycins) [reply saleor/order/migrations/tasks/saleor3_14.py:138]: btw. IMHO The best way would be `line.order_id`  (and fixing incorrect prefetch_related) + solid test on local instance.

File: saleor/checkout/utils.py [hunk 5]
```diff
@@ -954,3 +956,27 @@ def get_checkout_metadata(checkout: "Checkout"):
...
+    price: "TaxedMoney",
+    tax_rate: "Decimal",
+    prices_entered_with_tax: bool,
+):
+    """Apply taxes to undiscounted base price.
+
+    This function also prevents rounding difference between prices from tax-app and
+    local calculations based on tax_rate that might occur in orders without discounts.
+    """
+    price_to_compare = price.gross if prices_entered_with_tax else price.net
```
[the remark this answers is not in the document]
Comment (jakubkuc) [reply saleor/checkout/utils.py:972]: yes it does. Sometimes when there are no discounts tax app can return `tax_rate` and taxed `price` that are not "matching" (are rounded differently) so the prices from `TAX_APP` and saleors calculations were slightly different. This prevents those small differences.

File: saleor/plugins/manager.py (hunk 3 above)
Comment (jakubkuc) [reply saleor/plugins/manager.py:451]: It is a bug that was found during development of this PR... maybe the best way would be to fix this particular line in separate PR and port it to versions affected

Comment (maarcingebala) [reply saleor/plugins/manager.py:451]: To me it looks risky, I would remove and fix in a separate PR, with proper testing and QA only of this single case.

Comment (maarcingebala) [reply saleor/plugins/manager.py:451]: Also with an GH issue first that proves and shows why and when this calculation is incorrect.

**Comment by maarcingebala** [comment]
Closing in favor of https://github.com/saleor/saleor/pull/14095.
